File: C#/CrossCall/CrossCall.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using API_HubTool.Bindings;
using static API_HubTool.Bindings.API;

namespace CrossDemo
{
    class CrossCall
    {
        // ---------- add 远程调用 ----------
        private static int Add(int a, int b)
        {
            DataHnd send = API_Create_DataHnd("add");
            if (!API_WriteInt32(send, a) || !API_WriteInt32(send, b))
            {
                API_Free_DataHnd(send);
                return 0;
            }

            DataHnd result = API_Call("demo", send, 1000);
            API_Free_DataHnd(send);

            if (API_GetSize(result) == 0)
            {
                API_Free_DataHnd(result);
                return 0;
            }
            if (!API_ReadInt32(result, out int c))
                c = 0;
            API_Free_DataHnd(result);
            return c;
        }

        // ---------- inv_seri 远程调用 ----------
        private static string InvSeri()
        {
            // 构造输入数据
            byte b = 200;
            ushort w = 0x10;
            uint c = 0x2F;
            ulong u64 = 0x3F;
            string s = "hello world";
            float f = 3.14f;

            DataHnd send = API_Create_DataHnd("inv_seri");
            if (!API_WriteUInt8(send, b) ||
                !API_WriteUInt16(send, w) ||
                !API_WriteUInt32(send, c) ||
                !API_WriteUInt64(send, u64))
            {
                API_Free_DataHnd(send);
                return "写入基本类型失败";
            }
            // ⭐ 使用官方 API.WriteString（空终止符协议）
            if (!API_WriteString(send, s))
            {
                API_Free_DataHnd(send);
                return "写入字符串失败";
            }
            if (!API_WriteSingle(send, f))
            {
                API_Free_DataHnd(send);
                return "写入 float 失败";
            }

            DataHnd result = API_Call("demo", send, 1000);
            API_Free_DataHnd(send);

            if (API_GetSize(result) == 0)
            {
                API_Free_DataHnd(result);
                return "超时";
            }

            // 按逆序读取（使用官方 API.ReadString）
            if (!API_ReadSingle(result, out float f2))
            {
                API_Free_DataHnd(result);
                return "读取 float 失败";
            }
            string s2 = API_ReadString(result);    // ⭐ 自动读取到 #0
            if (!API_ReadUInt64(result, out ulong u64_2) ||
                !API_ReadUInt32(result, out uint c2) ||
                !API_ReadUInt16(result, out ushort w2) ||
                !API_ReadUInt8(result, out byte b2))
            {
                API_Free_DataHnd(result);
                return "读取基本类型失败";
            }
            API_Free_DataHnd(result);

            return $"接收逆序: [{b2}, {w2}, {c2}, {u64_2}, \"{s2}\", {f2}]  原始: [{b}, {w}, {c}, {u64}, \"{s}\", {f}]";
        }

        static void Main()
        {
            Console.WriteLine("=== Cross Call (Client) ===");

            // 连接服务端点（纯消费，不暴露 API）
            API_Reset_Prepare();
            int prep = API_Prepare_Client("ipc:cross", AppHnd.Null);
            if (prep == 0)
            {
                Console.WriteLine("API_Prepare_Client 失败");
                return;
            }

            if (API_Prepare_Done() != 1)
            {
                Console.WriteLine("API_Prepare_Done 失败，请检查控制台输出");
                API_shutdown();
                return;
            }

            Console.WriteLine("已连接到 ipc:cross，开始负载测试（10秒）");

            CancellationTokenSource cts = new CancellationTokenSource();
            cts.CancelAfter(TimeSpan.FromSeconds(10));

            // 使用 Parallel.For，检查外部 CancellationToken
            Parallel.For(0, 10, (i, state) =>
            {
                var rand = new Random(i + Environment.TickCount);
                while (!cts.Token.IsCancellationRequested)
                {
                    if (rand.Next(2) == 0)
                    {
                        int a = rand.Next(1, 1000);
                        int b = rand.Next(1, 1000);
                        int result = Add(a, b);
                        if (result != 0)
                            Console.WriteLine($"[Client {i}] add({a},{b}) = {result}");
                        else
                            Console.WriteLine($"[Client {i}] add({a},{b}) 超时或失败");
                    }
                    else
                    {
                        string msg = InvSeri();
                        Console.WriteLine($"[Client {i}] {msg}");
                    }
                    Thread.Sleep(50);
                }
            });

            Console.WriteLine("负载测试结束，按 Enter 退出...");
            Console.ReadLine();

            API_Exit_MainThread();
            API_shutdown();
        }
    }
}

File: C++/CrossDemo/CrossCall.cpp
// CrossCall.cpp - Client that calls 'add' and 'inv_seri' concurrently.
// It connects to the IPC endpoint "ipc:cross" and runs a 10-second load test.

#include "API_HubTool.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <random>
#include <cstdint>
#include <atomic>
#include <vector>

using namespace z_api_hub;

// ---------- Remote call: add ----------
static int32_t Add(int32_t a, int32_t b) {
    DataHandle param("add");
    param.writeInt32(a);
    param.writeInt32(b);

    TDataHnd rawResult = API_Call("demo", param.get(), 1000);
    if (!rawResult || API_GetSize(rawResult) == 0) {
        if (rawResult) API_Free_DataHnd(rawResult);
        return 0;
    }
    DataHandle result(rawResult, true);   // take ownership
    int32_t c = 0;
    if (!result.readInt32(&c)) c = 0;
    return c;
}

// ---------- Remote call: inv_seri ----------
static std::string InvSeri() {
    // Input data sequence
    uint8_t  b = 200;
    uint16_t w = 0x10;
    uint32_t c = 0x2F;
    uint64_t u64 = 0x3F;
    std::string s = "hello world";
    float    f = 3.14f;

    DataHandle param("inv_seri");
    param.writeUInt8(b);
    param.writeUInt16(w);
    param.writeUInt32(c);
    param.writeUInt64(u64);
    param.writeString(s);
    param.writeSingle(f);

    TDataHnd rawResult = API_Call("demo", param.get(), 1000);
    if (!rawResult || API_GetSize(rawResult) == 0) {
        if (rawResult) API_Free_DataHnd(rawResult);
        return "Timeout or error";
    }
    DataHandle result(rawResult, true);

    // Read back in reverse order
    float    f2;
    std::string s2;
    uint64_t u64_2;
    uint32_t c2;
    uint16_t w2;
    uint8_t  b2;

    if (!result.readSingle(&f2) ||
        !result.readString(&s2) ||
        !result.readUInt64(&u64_2) ||
        !result.readUInt32(&c2) ||
        !result.readUInt16(&w2) ||
        !result.readUInt8(&b2)) {
        return "Failed to read reply";
    }

    char buf[256];
    snprintf(buf, sizeof(buf),
        "Reply: [%d, %d, %d, %llu, \"%s\", %.2f]  Original: [%d, %d, %d, %llu, \"%s\", %.2f]",
        (int)b2, w2, c2, (unsigned long long)u64_2, s2.c_str(), f2,
        (int)b, w, c, (unsigned long long)u64, s.c_str(), f);
    return std::string(buf);
}

int main() {
    std::cout << "=== Cross Call (Client) ===" << std::endl;

    // Load library (RAII)
    LibraryLoader loader;

    // Connect as pure consumer (do not expose any API)
    resetPrepare();
    int prep = API_Prepare_Client("ipc:cross", nullptr);
    if (prep == 0) {
        std::cerr << "API_Prepare_Client failed." << std::endl;
        return 1;
    }

    if (API_Prepare_Done() != 1) {
        std::cerr << "API_Prepare_Done failed. Check console output." << std::endl;
        API_shutdown();
        return 1;
    }

    std::cout << "Connected to ipc:cross. Starting 10-second load test..." << std::endl;

    const int THREADS = 10;
    const int DURATION_SEC = 10;
    std::atomic<bool> stopFlag(false);
    std::vector<std::thread> threads;

    // Launch worker threads
    for (int i = 0; i < THREADS; ++i) {
        threads.emplace_back([i, &stopFlag]() {
            std::mt19937 rng(i + std::random_device{}());
            std::uniform_int_distribution<int> dist(0, 1);
            std::uniform_int_distribution<int> numDist(1, 1000);

            while (!stopFlag.load()) {
                if (dist(rng) == 0) {
                    int a = numDist(rng);
                    int b = numDist(rng);
                    int result = Add(a, b);
                    if (result != 0)
                        std::cout << "[Client " << i << "] add(" << a << "," << b << ") = " << result << std::endl;
                    else
                        std::cout << "[Client " << i << "] add(" << a << "," << b << ") timed out or failed." << std::endl;
                }
                else {
                    std::string msg = InvSeri();
                    std::cout << "[Client " << i << "] " << msg << std::endl;
                }
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }
            });
    }

    // Run for DURATION_SEC seconds
    std::this_thread::sleep_for(std::chrono::seconds(DURATION_SEC));
    stopFlag.store(true);

    // Join all threads
    for (auto& t : threads) {
        if (t.joinable()) t.join();
    }

    std::cout << "Load test finished. Press Enter to exit..." << std::endl;
    std::cin.get();

    API_Exit_MainThread();
    API_shutdown();
    return 0;
}